internal/proxy/powerscale_handler.go

```diff
@@ -216,40 +281,116 @@ func (h *PowerScaleHandler) writeError(w http.ResponseWriter, msg string, code i
+	sessionStatusResp, err := client.Do(sessionStatusReq)
+	if err != nil {
+		return fmt.Errorf("Error requesting session cookie status for PowerScale %v: %e", v.Endpoint, err)
+	}
+	sessionStatusRespBody, err := ioutil.ReadAll(sessionStatusResp.Body)
+	if err != nil {
+		return fmt.Errorf("Error reading session status response body: %e", err)
+	}
+	h.log.Debugf("get session status response: (%v) %v", sessionStatusResp.StatusCode, string(sessionStatusRespBody))
+
+	// If not valid, get a new session cookie
+	if sessionStatusResp.StatusCode == http.StatusUnauthorized {
+		h.log.Info("Authintication session is expired. Requesting a new session...")
+		type newSessionRequestBody struct {
+			Username string   `json:"username"`
+			Password string   `json:"password"`
+			Services []string `json:"services"`
+		}
+		req := newSessionRequestBody{
+			Username: v.User,
+			Password: v.Password,
+			Services: []string{"platform", "namespace"},
+		}
+		reqBody, err := json.Marshal(req)
+		if err != nil {
+			return fmt.Errorf("Failed to marshal session request body: %e", err)
+		}
+		h.log.Debugf("New session request body: %v", string(reqBody))
+		newSessionResp, err := http.Post(v.Endpoint+"/session/1/session", "application/json", bytes.NewBuffer(reqBody))
+		if err != nil {
+			return fmt.Errorf("Error requesting new session: %e", err)
+		}
+		defer newSessionResp.Body.Close()
+
+		respBody, err := ioutil.ReadAll(newSessionResp.Body)
+		if err != nil {
+			return fmt.Errorf("reading response body from new session request: %e", err)
+		}
+		if newSessionResp.StatusCode != http.StatusCreated {
+			return fmt.Errorf("in response when requesting session token: %v", string(respBody))
+		}
+		h.log.Debugf("New session response: (%v) %v", newSessionResp.StatusCode, string(respBody))
+
+		headerRes := strings.Join(newSessionResp.Header.Values("Set-Cookie"), " ")
+
+		startIndex, endIndex, matchStrLen := FetchValueIndexForKey(headerRes, "isisessid=", ";")
+		v.sessionCookie = headerRes[startIndex : startIndex+matchStrLen+endIndex]
+		if startIndex < 0 || endIndex < 0 {
+			return fmt.Errorf("Could not extract isisessid from new session response: %v", headerRes)
+		}
+
+		startIndex, endIndex, matchStrLen = FetchValueIndexForKey(headerRes, "isicsrf=", ";")
+		v.csrfToken = headerRes[startIndex+matchStrLen : startIndex+matchStrLen+endIndex]
+		if startIndex < 0 || endIndex < 0 {
+			h.log.Errorf("Could not extract isisessid from new session response: %v", headerRes)
+		}
+	}
+
+	// Add the session cookie to the request's headers
+	r.Header.Add("Cookie", v.sessionCookie)
+	h.log.Debugf("added session cookie to request header: %v", v.sessionCookie)
+	r.Header.Add("X-CSRF-Token", v.csrfToken)
+	h.log.Debugf("added CSRF token to request header: %v", v.csrfToken)
+
+	// Add referrer header
+	r.Header.Add("Referer", v.Endpoint)
+	return nil
+}
+
+func FetchValueIndexForKey(l string, match string, sep string) (int, int, int) {
```

Comment:
According to the linting check this should probably start with a lowercase letter:
`internal/proxy/powerscale_handler.go:357:1: exported function FetchValueIndexForKey should have comment or be unexported`
(my bad)

Reply:
Youll have to update the one spot in this file where this function is called too

Reply:
Fixed.